Source/Private/ARTOSReachability.m

```diff
@@ -38,10 +32,24 @@ - (instancetype)initWithLogger:(ARTLog *)logger queue:(dispatch_queue_t)queue {
 - (void)listenForHost:(NSString *)host callback:(void (^)(BOOL))callback {
     [self off];
     _host = host;
-    _callback = callback;
 
+    __weak ARTOSReachability *weakSelf = self;
+    void (^callbackBlock)(SCNetworkReachabilityFlags) = ^(SCNetworkReachabilityFlags flags) {
+        ARTOSReachability *strongSelf = weakSelf;
+        if (strongSelf) {
+            BOOL reachable = (flags & kSCNetworkReachabilityFlagsReachable) != 0;
+            [strongSelf->_logger info:@"Reachability: host %@ is reachable: %@", strongSelf->_host, reachable ? @"true" : @"false"];
+            dispatch_async(strongSelf->_queue, ^{
+                if (callback) {
+                    callback(reachable);
+                }
+            });
+        }
+    };
+    
     _reachabilityRef = SCNetworkReachabilityCreateWithName(NULL, [host UTF8String]);
-    SCNetworkReachabilityContext context = {0, (__bridge void *)(self), NULL, NULL, NULL};
+    
+    SCNetworkReachabilityContext context = { .version = 0, .info = (void *)CFBridgingRetain(callbackBlock), .release = CFRelease };
```

Comment:
I don't really understand how this memory management works. Do you know whether `SCNetworkReachabilityContext` expects to be passed an `info` that's already been retained? On a related note, under what circumstances will it call its `.release` if you don't also pass a `.retain` callback?

Here are the retains and releases performed during a run of `RealtimeClientTests` as this stands (I added logging [here](https://github.com/ably/ably-cocoa/commit/0cee3217279332e18f7f6d27fcf313aac359957b) and then piped the result through `egrep 'Retain|Release' | cut -f 4,5 -d ' '`):

```
0x600003a6d650 CFBridgingRetain
0x600003a74b40 CFBridgingRetain
0x600003a6e190 CFBridgingRetain
0x600003a6e190 myRelease
0x600003a6ae50 CFBridgingRetain
0x600003a6ed60 CFBridgingRetain
0x600003a6ae50 myRelease
0x600003a6ed60 myRelease
0x600003a7c5d0 CFBridgingRetain
0x600003a7c5d0 myRelease
0x600003a86280 CFBridgingRetain
0x600003a7cab0 CFBridgingRetain
0x600003a86280 myRelease
0x600003a7cab0 myRelease
0x600003a79a10 CFBridgingRetain
0x600003a79a10 myRelease
0x600003a6e610 CFBridgingRetain
0x600003a8d080 CFBridgingRetain
0x600003a8d080 myRelease
0x600003a6e610 myRelease
0x600003a799b0 CFBridgingRetain
0x600003a799b0 myRelease
0x600003a79a40 CFBridgingRetain
0x600003a6dbf0 CFBridgingRetain
0x600003a79a40 myRelease
0x600003a6dbf0 myRelease
0x600003a87390 CFBridgingRetain
0x600003a87390 myRelease
0x600003a79e30 CFBridgingRetain
0x600003a79e30 myRelease
0x600003a7b2d0 CFBridgingRetain
0x600003a7b2d0 myRelease
0x600003a7edc0 CFBridgingRetain
0x600003a7edc0 myRelease
0x600003a71c50 CFBridgingRetain
0x600003a71c50 myRelease
0x600003a9d200 CFBridgingRetain
0x600003a9d200 myRelease
0x600003a913b0 CFBridgingRetain
0x600003a913b0 myRelease
0x600003a7f690 CFBridgingRetain
0x600003a7f690 myRelease
0x600003a8a280 CFBridgingRetain
0x600003a8a280 myRelease
0x600003a639f0 CFBridgingRetain
0x600003a639f0 myRelease
0x600003aac540 CFBridgingRetain
0x600003aac540 myRelease
0x600003aaee80 CFBridgingRetain
0x600003aaee80 myRelease
0x600003aac990 CFBridgingRetain
0x600003a8f3f0 CFBridgingRetain
0x600003a8f3f0 myRelease
0x600003ab4900 CFBridgingRetain
0x600003ab4900 myRelease
0x600003ab0ba0 CFBridgingRetain
0x600003a5f960 CFBridgingRetain
0x600003a9dc80 CFBridgingRetain
0x600003a9dc80 myRelease
0x600003a7e7f0 CFBridgingRetain
0x600003ab2880 CFBridgingRetain
0x600003a833f0 CFBridgingRetain
0x600003a9d0e0 CFBridgingRetain
0x600003ab3e40 CFBridgingRetain
0x600003a9c7b0 CFBridgingRetain
0x600003ac40c0 CFBridgingRetain
0x600003a74270 CFBridgingRetain
0x600003ac03f0 CFBridgingRetain
0x600003ac0600 CFBridgingRetain
0x600003ac1950 CFBridgingRetain
```

You can see that whilst some of the calls are balanced, some of them are not. I wonder what's going on here? I also tried not retaining the `callbackBlock` up front and instead passing a `.retain` callback ([here](https://github.com/ably/ably-cocoa/commit/1e13c1d16131dcb0ba23933f5a51350b79abb45f)) and that gave this:

```
0x60000184e100 myRetain
0x60000184e100 myRetain
0x60000184e100 myRelease
0x600001831c80 myRetain
0x600001864f00 myRetain
0x600001864f00 myRetain
0x600001864f00 myRelease
0x600001864f00 myRelease
0x600001843090 myRetain
0x600001849710 myRetain
0x600001843090 myRetain
0x600001843090 myRelease
0x600001849710 myRetain
0x600001849710 myRelease
0x600001843090 myRelease
0x600001849710 myRelease
0x600001855890 myRetain
0x600001855890 myRelease
0x6000018bc1b0 myRetain
0x6000018bc1b0 myRetain
0x6000018bc1b0 myRelease
0x6000018b9950 myRetain
0x6000018bc1b0 myRelease
0x6000018b9950 myRetain
0x6000018b9950 myRelease
0x6000018b9950 myRelease
0x60000184f150 myRetain
0x60000184f150 myRetain
0x60000184f150 myRelease
0x60000184f150 myRelease
0x600001852f40 myRetain
0x600001852f40 myRetain
0x600001852f40 myRelease
0x6000018ba610 myRetain
0x6000018ba610 myRetain
0x6000018ba610 myRelease
0x6000018ba610 myRelease
0x600001852f40 myRelease
0x6000018bd200 myRetain
0x6000018bd200 myRetain
0x6000018bd200 myRelease
0x6000018bd200 myRetain
0x6000018bd200 myRelease
0x6000018bd200 myRelease
0x600001849710 myRetain
0x600001854f90 myRetain
0x600001849710 myRelease
0x600001854f90 myRelease
0x600001864cc0 myRetain
0x600001864cc0 myRelease
0x600001853630 myRetain
0x600001853630 myRetain
0x600001853630 myRelease
0x600001853630 myRelease
0x60000184fae0 myRetain
0x60000184fae0 myRetain
0x60000184fae0 myRelease
0x60000184fae0 myRelease
0x6000018bf330 myRetain
0x6000018bf330 myRetain
0x6000018bf330 myRelease
0x6000018bf330 myRelease
0x6000018b5a70 myRetain
0x6000018b5a70 myRetain
0x6000018b5a70 myRelease
0x6000018b5a70 myRelease
0x600001867b10 myRetain
0x600001867b10 myRetain
0x600001867b10 myRelease
0x600001867b10 myRelease
0x6000018a0240 myRetain
0x6000018a0240 myRetain
0x6000018a0240 myRelease
0x6000018a0240 myRelease
0x6000018a48d0 myRetain
0x6000018a48d0 myRetain
0x6000018a48d0 myRelease
0x6000018a48d0 myRelease
0x6000018ada40 myRetain
0x6000018ada40 myRetain
0x6000018ada40 myRelease
0x6000018ada40 myRelease
0x6000018aa550 myRetain
0x6000018aa550 myRetain
0x6000018aa550 myRelease
0x6000018aa550 myRelease
0x6000018a26d0 myRetain
0x6000018a26d0 myRetain
0x6000018a26d0 myRelease
0x6000018a26d0 myRelease
0x60000189e9a0 myRetain
0x60000189e9a0 myRetain
0x60000189e9a0 myRelease
0x60000189e9a0 myRelease
0x600001895080 myRetain
0x60000188a2b0 myRetain
0x60000188a2b0 myRetain
0x60000188a2b0 myRelease
0x60000188a2b0 myRelease
0x60000188de60 myRetain
0x60000188de60 myRetain
0x60000188de60 myRelease
0x60000188de60 myRelease
0x60000188e0a0 myRetain
0x60000188e0a0 myRetain
0x60000188e0a0 myRelease
0x600001857a50 myRetain
0x60000188a700 myRetain
0x60000188a700 myRetain
0x60000188a700 myRelease
0x60000188a700 myRelease
0x6000018ac6c0 myRetain
0x6000018b39f0 myRetain
0x6000018b39f0 myRetain
0x6000018b39f0 myRelease
0x6000018af870 myRetain
0x6000018a46c0 myRetain
0x6000018a46c0 myRetain
0x6000018a46c0 myRelease
0x6000018afc60 myRetain
0x60000188bc90 myRetain
0x60000188bc90 myRetain
0x60000188bc90 myRelease
0x6000018a42d0 myRetain
0x6000018a42d0 myRetain
0x6000018a42d0 myRelease
0x600001885110 myRetain
0x600001885110 myRetain
0x600001885110 myRelease
0x60000188dec0 myRetain
0x60000188dec0 myRetain
0x60000188dec0 myRelease
0x6000018f00f0 myRetain
0x6000018f00f0 myRetain
0x6000018f00f0 myRelease
0x6000018f4cc0 myRetain
0x6000018f4cc0 myRetain
0x6000018f4cc0 myRelease
```

This is just as confusing — some calls are balanced, some are not.

Do you have any thoughts on what's going on here?

Reply:
Mike Ash normally knows what he's talking about, so let's go with it 😀 Can you add a code comment explaining where you got this strategy from? https://www.mikeash.com/pyblog/friday-qa-2013-06-14-reachability.html

Reply:
https://github.com/ably/ably-cocoa/pull/1565/commits/036be28852b3ef1b3a112aea7052c463c5b8792a Some interesting comments about the behavior of Reachability there 🧐

Reply:
I'm still uneasy about this one, actually. Given the potential consequences of getting memory management wrong (e.g. over-release leading to app crash) I want to be sure we're careful about this and don't just take somebody else's solution, even if we believe them trustworthy. Just by symmetry, if nothing else, it seems really weird that you'd tell `SCNetworkReachability` how to release an object but now how to retain it. It shouldn't be releasing anything unless one of the following is true:

- the release balances some corresponding action that it has performed itself (i.e. a retain);
- the release balances some corresponding action that was performed externally, about which a guarantee is made (e.g. that ownership has been in some way transferred)

Take a look at these examples of third-party libraries that wrap `SCNetworkReachability` (the first couple to come up when I Googled "swift reachability"):

- https://github.com/ashleymills/Reachability.swift/blob/a81b7367f2c46875f29577e03a60c39cdfad0c8d/Sources/Reachability.swift#L190-L208
- https://github.com/vsanthanam/NetworkReachability/blob/918943ba196327e1256a8f2b7e8b922651488607/Sources/NetworkReachability/API/ReachabilityMonitor/ReachabilityMonitor.swift#L529-L547

You can see they both pass `retain` and `release` callbacks.

Please could you do some more investigation to find out about how memory management is meant to work with `SCNetworkReachability`? I think Apple made a well-known example project once, and there might also be some documentation.

Reply:
https://github.com/ably/ably-cocoa/pull/1565/commits/b72f778c5934b5a24243b52dd5fed409ce8f34a7